nearcore/src/lib.rs

```diff
@@ -74,17 +71,89 @@ pub fn get_default_home() -> PathBuf {
+    let path = db_checkpoint_path(path, near_config);
+    if path.exists() {
+        Some(path)
+    } else {
+        None
+    }
+}
+
+/// Creates a consistent DB checkpoint and returns its path.
+/// By default it creates checkpoints in the DB directory, but can be overridden by the config.
+fn create_db_checkpoint(path: &Path, near_config: &NearConfig) -> Result<PathBuf, anyhow::Error> {
+    let checkpoint_path = db_checkpoint_path(path, near_config);
+
+    let db = RocksDB::new(path)?;
+    let checkpoint = db.checkpoint()?;
+    info!(target:"near","Creating a DB snapshot in '{}'",checkpoint_path.display());
+    checkpoint.create_checkpoint(&checkpoint_path)?;
+    info!(target:"near","Created a DB snapshot in '{}'",checkpoint_path.display());
+
+    Ok(checkpoint_path)
+}
+
 /// Function checks current version of the database and applies migrations to the database.
 pub fn apply_store_migrations(path: &Path, near_config: &NearConfig) {
     let db_version = get_store_version(path);
     if db_version > near_primitives::version::DB_VERSION {
         error!(target: "near", "DB version {} is created by a newer version of neard, please update neard or delete data", db_version);
         std::process::exit(1);
     }
+
     if db_version == near_primitives::version::DB_VERSION {
         return;
     }
 
+    // Before attemping a DB migration check if the current DB is corrupted. We assume that
+    // checkpoints get created before each migration attempt and get removed after successful
+    // attempts.
+    if near_config.config.use_checkpoints_for_db_migration {
+        match check_db_checkpoint_exists(path, near_config) {
+            Some(existing_checkpoint_path) => {
+                panic!("Detected an existing DB migration checkpoint: '{}'. Probably a DB migration got interrupted and your DB is corrupted. Please replace the contents of '{}' with data from that checkpoint, delete the checkpoint and try again.",existing_checkpoint_path.display(), path.display());
+            }
+            None => {}
+        }
+    }
+
+    // Before starting a DB migration, create a consistent snapshot of the database. If a migration
+    // fails, it can be used to quickly restore the database to its original state.
+    let checkpoint_path = if near_config.config.use_checkpoints_for_db_migration {
+        match create_db_checkpoint(path, near_config) {
+            Ok(checkpoint_path) => {
+                info!(target: "near", "Created a DB checkpoint before a DB migration: '{}'. Please recover from this checkpoint if the migration gets interrupted.", checkpoint_path.display());
+                Some(checkpoint_path)
+            }
+            Err(err) => {
+                panic!("Failed to create a DB checkpoint before a DB migration: {}. If this persists, you can disable checkpoints in `config.json`:\n```\"use_checkpoints_for_db_migration\": false\n```", err);
+            }
+        }
+    } else {
+        None
+    };
```

Comment:
Nit, but to subsequent blocks which start with `if near_config.config.use_checkpoints_for_db_migration` feel like something which can be compressed. 

Why do we need separate "exits/create" API? Can we have just a single `create_db_checkpoint` which fails if there's an existing checkpoints? This'll avoid TOCTOU-prone API.